### ISSUE TITLE
Mute Audio

### ISSUE THREAD

**Issue by wheezy1749** [ladaapp/lada#3]
My own fault. But please mute audio by default and add a toggle to the GUI. Just had a loud slurping blowjob noise just blast through my speakers at night waiting for a video to load. Lol.

Edit: Also, great repo. Thanks for sharing. Have been making some changes to the cli. Prefer the output as image files so added that option and improved the progress bars. I had to look into the code to see why it kept pausing. Makes sense but added a progress indication during that time. Can do a PR if you want them. Just some quick changes for my use case. Will probably be messing around more with the code.

Edit2: Working on an option for larger videos that pre-samples regions of the video with the detection model at a given frame interval. The goal is to only run the mosaic detection on every Nth frame first (saving the model info to a temporary database so we don't have to re-run these sampled frames through the model later when we actually clean them). Then determine regions of the video that contain mosaics within the N frame density. Would greatly improve the time it takes to process a large video. But would have a downside of missing mosaics that appear for less than N frames. But passing the detection model to every single frame of the video (like it does now) is overkill I think. Would set it as a CLI option and keep your default behavior intact.

Let me know if you're interested in any of these ideas. Otherwise I'll just fork and you can do your own thing.

**Comment by ladaapp** [comment]
On the first point (1) -> lol indeed
Having mute (and volume control slider) functionality would be nice to have but I wouldn't set the default to mute (cannot think of any other video/audio tool/player that mutes on default). Until then you could mute via Volume Control/pavucontrol if you're on linux.

On output to images files (2):
What's your use case for this? I guess you want to do some post-processing afterwards? :thinking: 

On only sampling every nth frame in the mosaic detection model (3):
Hmm, if I understand it correctly then I'm probably not interested in this given the additional complexity and downsides of missing frames / incorrect region estimates for the frames that will not run though the detection model.
An optimization for skipping over non nsfw mosaic scenes maybe useful (based on a subset of frames to determine if we should process it properly) but I'm not sure if its worth it. I didn't do proper profiling but as we'd still have to decode and re-encode these all these frames anyway to produce the final video I'm not sure we'd save that much time. The base model (yolo v11 seg) is quite fast [performance metrics](https://docs.ultralytics.com/models/yolo11/#performance-metrics )

Don't quote me on this, just a guess: I'd assume the bigger time save would be to optimize the video decoding / frame feeding to the detection model. Currently this is done frame-by-frame using PyAv to decode. You could try  to integrate [TorchCodec](https://github.com/pytorch/torchcodec) and feed frame batches to the Yolo model potentially using GPU decoding if available given hardware/input files.

**Comment by ladaapp** [comment]
FYI, ran the profiler on it:

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   149916 1013.204    0.007 1013.204    0.007 video_utils.py:234(frames)  <-- video decoding
    74957  563.562    0.008  618.431    0.008 video_utils.py:494(write)  <-- video encoding
   149914    2.371    0.000  613.493    0.004 predictor.py:219(stream_inference)  <-- mosaic detection
      757    0.159    0.000  351.196    0.464 restored_mosaic_frames_generator.py:61(restore_clip)  <-- mosaic restoration
```
So biggest impact should have: 
1) Get rid of the double decoding (it currently decodes it once to feed to mosaic detection and another time for restoration)
2) Batch feed frames to mosaic detection model (clip/scene creation is done frame-by-frame still so this should be done in the frame generator and cached so subsequent calls to next() will be served from cache until we need to run another batch)
3) async processing for the video encoding part as it should fully IO/CPU bound so we could use GPU to run restoration/detection.
4) try if TorchCodec speeds up video decoding step

@wheezy1749 Let me know if you're interesting in any of this then I would create an issue for it

**Comment by wheezy1749** [comment]
I messed around with some of these. Biggest issue right now is the code is pretty fragmented. It runs well. But its hard to make changes without getting stuck in some issue. I'm working on some re-factoring since I'm not seeing the performance increase I'd like from removing the double decoding. Not anything I'd expect you to pull. Just messing around to see what I can get.

If you don't mind create branch permissions for me we can discuss what I come up with. Would save my sanity as I keep forgetting I'm on a fork and trying to fetch latest. lol

I'd ask/mention a couple things:

(1) There are some extra features in the code like mosaic cleaning (which says you don't recommend). preserving/not preserving scale. Right now while I'm testing things I'm just removing these options to focus on the standard removal as I refactor some stuff. Your model is really great though.

(2) I'm entirely ignoring gui stuff since that's not important to me. (despite what this "issue" started from" lol)  So keep that in mind.

(3) I'm trying to "classify" the code you have so different features could be added more easily. Basically, I think the code needs a bit of a refactoring before I tackle the optimizations you mentioned. My goal right now is to learn from what you made and try to get to the same or better performance you while splitting it into more easily digestible parts.

**Comment by ladaapp** [comment]
Cool, have fun. Hope you can find some improvements :rocket: .
I'm not planning to do any refactoring or bigger changes soon, so there shouldn't be conflicts with the stuff you're working on.
Just fixing bugs for issues I encounter and potentially some GUI adjustments to make it a bit less clunky / more robust.

Here are some remarks on the two topics you mentioned to give you bit more background on it:

### Mosaic Cleaning
Yeah mosaic cleaning is not worth it at least in it's current state. I tried a view approaches but it turned out that increasing the dataset with more diverse types of degradations and mosaic types and training the model longer with this raw/degraded mosaic clips is more robust than cleaning them up-front. At least for the current version of the restoration model, looks like it has enough capacity to learn it.
Maybe it has some merit when using another model architecture or training a smaller version of the current model (I don't intend to work on this, just saying) hence why I still kept it.

### Preserving / Not preserving scale
It's not that big of a difference in practice. I started with not preserving scale (zooming each frame to fit target dimension) as that is what DeepMosaics did. Training with clips that preserve the sale (pad each frame crop if necessary) showed better quality though so I switched and that is what the current version is trained with and what we do when inferencing.
Type of padding didn't matter AFAIK hence I sticked to zero-padding. (At least when training on a smaller subset)
Depending on the data / mosaic region size / how much movement there is between each frames (and framerate) and how many frames we consider for a Clip there can be quite a lot of padding and need to downsize the clip to fit target dimension 256x256.
In practice, I didn't see much of a problem here but there may be some room of potential to tune this scaling/cropping/padding params.

**Comment by ladaapp** [comment]
Added mute functionality:
* mute/unmute button
* mute/unmute shortcut (`m`)
* default is not muted when you open a file but you can overwrite this in the sidebar

Will land in the next release/flathub.

@wheezy1749
Regarding the other things discussed here, feel free to open a new issue for further discussion. Closing this ticket now
